aip/0160.md

```diff
@@ -0,0 +1,167 @@
+something reasonably similar), it is desirable to filter over the collection
+and only return results that the user is interested in.
+
+It is tempting to define a structure to handle the precise filtering needs for
+each API. However, filtering requirements evolve frequently, and therefore it
+is prudent to use a string field with a structured syntax accessible to a
+non-technical audience. This allows updates to be able to be made
+transparently, without waiting for UI or client updates.
+
+**Note:** Because list filters are intended for a potentially non-technical
+audience, they sometimes borrow from patterns of colloquial speech rather than
+common patterns found in code.
+
+## Guidance
+
+APIs **may** provide filtering to users on `List` methods (or similar methods
+to query a collection, such as `Search`). If they choose to do so, they
+**must** follow the common specification for filters discussed here. The syntax
+is formally defined in the [EBNF grammar][].
+
+**Note:** Google provides implementations for the filtering language in both
+[C++][cel-cpp] and [Go][cel-go] as a discrete component of [CEL][]. The use of
+these implementations is encouraged.
+
+### Expressions
+
+Expressions are literal values in the filtering syntax ("42", "Victor Hugo").
+Expressions can be pieced together with operators to make other expressions.
+
+### Logical Operators
+
+Filtering implementations **must** provide the binary operators `AND` and `OR`,
+as well as a whitespace operator:
+
+| Operator | Example       | Meaning                                |
+| -------- | ------------- | -------------------------------------- |
+| `AND`    | `a AND b`     | True if `a` and `b` are true.          |
+| `OR`     | `a OR b OR c` | True if any of `a`, `b`, `c` are true. |
+
+The whitespace operator (example: `a b c`) represents a fuzzy variant of `AND`,
+which is suitable for result ranking and term rewrite.
+
+**Note:** To match common patterns of speech, the `OR` operator has higher
+precedence than `AND`, unlike what is found in most programming languages. The
+expression `a AND b OR c` evaluates: `a AND (b OR c)`. API documentation and
+examples **should** encourage the use of explicit parentheses to avoid
+confusion.
+
+### Negation Operators
+
+Filtering implementations **must** provide the unary operators `NOT` and `-`.
+These are used interchangeably.
+
+| Operator | Example | Meaning                  |
+| -------- | ------- | ------------------------ |
+| `NOT`    | `NOT a` | True if `a` is not true. |
+| `-`      | `-a`    | True if `a` is not true. |
+
+### Comparison Operators
+
+Filtering implementations **must** provide the binary comparison operators `=`,
+`!=`, `<`, `>`, `<=`, and `>=`.
+
+| Operator | Example      | Meaning                                         |
+| -------- | ------------ | ----------------------------------------------- |
+| `=`      | `a = true`   | True if `a` is true.                            |
```

Comment:
These are all examples against constants. Can I specify a filter of `address1 = address2` or similar to match one field against another?

Reply:
I am having trouble understanding how someone would do that in a filter.

Reply:
Exactly as address1=address2, rather than address1='some value'. It might make more sense in other queries, e.g. order_count = delivery_count

Reply:
Need to defer to @TristonianJones on this one, as I am not sure if this is supported.

Reply:
Filtering is designed to treat the right-hand side of the expression as a literal rather than as a field reference. `address1 = address2` would be interpreted by most implementations as `address1 = "address2"`.

Reply:
Okay, that was kind of what I expected.